From 8eb4129d3f90172d43074faf57c3180d6dd0fef5 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Thu, 22 Jan 2026 16:43:30 +0100
Subject: [PATCH] use timeval for last_update

instead of integer unix timestamps, use timeval to get sub second precision.
---
 lib/nsutils.c                | 25 ++++++++++++++-
 lib/nsutils.h                | 38 +++++++++++++++++++++++
 src/naemon/broker.c          | 60 ++++++++++++++++--------------------
 src/naemon/checks_host.c     | 40 +++++++++++-------------
 src/naemon/checks_service.c  | 59 ++++++++++++++++++-----------------
 src/naemon/commands.c        | 25 +++++----------
 src/naemon/logging.c         |  2 +-
 src/naemon/notifications.c   | 44 ++++++++++++--------------
 src/naemon/objects_host.h    |  2 +-
 src/naemon/objects_service.h |  2 +-
 src/naemon/sehandlers.c      | 16 +++++-----
 src/naemon/shared.c          | 10 +++---
 src/naemon/workers.c         | 18 +----------
 src/naemon/xrddefault.c      |  8 ++---
 src/naemon/xsddefault.c      |  6 ++--
 src/worker/worker.c          |  4 +--
 16 files changed, 189 insertions(+), 170 deletions(-)

diff --git a/lib/nsutils.c b/lib/nsutils.c
index e3b600ca0..e814285a6 100644
--- a/lib/nsutils.c
+++ b/lib/nsutils.c
@@ -1,6 +1,7 @@
 #include "lnae-utils.h"
 #include "nsutils.h"
 #include <sys/types.h>
+#include <sys/time.h>
 #include <stdarg.h>
 #include <stdio.h>
 
@@ -67,6 +68,28 @@ float tv_delta_f(const struct timeval *start, const struct timeval *stop)
 	return ret;
 }
 
+/* format duration seconds into human readable string */
+const char* tv_str(struct timeval *tv) {
+	return (char *)mkstr("%lu.%06lu", tv->tv_sec, tv->tv_usec);
+}
+
+ /* Convert string to timeval */
+int str2timeval(char *str, struct timeval *tv)
+{
+	char *ptr, *ptr2;
+
+	tv->tv_sec = strtoul(str, &ptr, 10);
+	if (ptr == str) {
+		tv->tv_sec = tv->tv_usec = 0;
+		return -1;
+	}
+	if (*ptr == '.' || *ptr == ',') {
+		ptr2 = ptr + 1;
+		tv->tv_usec = strtoul(ptr2, &ptr, 10);
+	}
+	return 0;
+}
+
 #define MKSTR_BUFS 256 /* should be plenty */
 const char *mkstr(const char *fmt, ...)
 {
@@ -111,4 +134,4 @@ void close_standard_fds(void)
 	open("/dev/null", O_WRONLY);
 
 	return;
-}
\ No newline at end of file
+}
diff --git a/lib/nsutils.h b/lib/nsutils.h
index 38406025c..907a9e8ec 100644
--- a/lib/nsutils.h
+++ b/lib/nsutils.h
@@ -6,6 +6,7 @@
 #endif
 
 #include <sys/types.h>
+#include <sys/time.h>
 #include <fcntl.h>
 
 NAGIOS_BEGIN_DECL
@@ -123,6 +124,43 @@ extern int tv_delta_msec(const struct timeval *start, const struct timeval *stop
  */
 extern float tv_delta_f(const struct timeval *start, const struct timeval *stop);
 
+/**
+ * clone source timestamp to destination timeval
+ * @param tv1 Destination timeval
+ * @param tv2 Source timeval
+ * @return nothing
+ */
+static inline void tv_clone(struct timeval *dst, struct timeval *src)
+{
+	dst->tv_sec = src->tv_sec;
+	dst->tv_usec = src->tv_usec;
+}
+
+/**
+ * set timestamp to target timeval
+ * @param tv Target timeval
+ * @return nothing
+ */
+static inline void tv_set(struct timeval *timestamp)
+{
+	gettimeofday(timestamp, NULL);
+}
+
+/**
+ * Convert timeval to str
+ * @param tv Source timeval
+ * @return A pointer to the formatted string on success.
+ */
+const char* tv_str(struct timeval *tv);
+
+/**
+ * Convert string to timeval
+ * @param str The timeval string (sec.usec)
+ * @param tv The target timeval
+ * @return 0 on success, -1 on errors
+ */
+extern int str2timeval(char *str, struct timeval *tv);
+
 /**
  * close and reopen stdin, stdout and stderr to /dev/null
  */
diff --git a/src/naemon/broker.c b/src/naemon/broker.c
index 593b7b78b..460f10731 100644
--- a/src/naemon/broker.c
+++ b/src/naemon/broker.c
@@ -19,12 +19,6 @@ struct kvvec *get_global_store(void)
 	return &global_store;
 }
 
-/* gets timestamp for use by broker */
-static inline void get_broker_timestamp(struct timeval *timestamp)
-{
-	gettimeofday(timestamp, NULL);
-}
-
 /******************************************************************/
 /************************* EVENT FUNCTIONS ************************/
 /******************************************************************/
@@ -43,7 +37,7 @@ void broker_program_state(int type, int flags, int attr)
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	/* make callbacks */
 	neb_make_callbacks(NEBCALLBACK_PROCESS_DATA, (void *)&ds);
@@ -64,7 +58,7 @@ void broker_log_data(int type, int flags, int attr, char *data, unsigned long da
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.entry_time = entry_time;
 	ds.data_type = data_type;
@@ -92,7 +86,7 @@ void broker_system_command(int type, int flags, int attr, struct timeval start_t
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.start_time = start_time;
 	ds.end_time = end_time;
@@ -138,7 +132,7 @@ int broker_event_handler(int type, int flags, int attr, int eventhandler_type, v
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.eventhandler_type = eventhandler_type;
 	if (eventhandler_type == SERVICE_EVENTHANDLER || eventhandler_type == GLOBAL_SERVICE_EVENTHANDLER) {
@@ -199,7 +193,7 @@ int broker_host_check(int type, int flags, int attr, host *hst, int check_type,
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.host_name = hst->name;
 	ds.object_ptr = (void *)hst;
@@ -259,7 +253,7 @@ int broker_service_check(int type, int flags, int attr, service *svc, int check_
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.host_name = svc->host_name;
 	ds.service_description = svc->description;
@@ -306,7 +300,7 @@ void broker_comment_data(int type, int flags, int attr, int comment_type, int en
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.comment_type = comment_type;
 	ds.entry_type = entry_type;
@@ -341,7 +335,7 @@ void broker_downtime_data(int type, int flags, int attr, int downtime_type, char
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.downtime_type = downtime_type;
 	ds.host_name = host_name;
@@ -381,7 +375,7 @@ void broker_flapping_data(int type, int flags, int attr, int flapping_type, void
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.flapping_type = flapping_type;
 	if (flapping_type == SERVICE_FLAPPING) {
@@ -419,7 +413,7 @@ void broker_program_status(int type, int flags, int attr)
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.program_start = program_start;
 	ds.pid = nagios_pid;
@@ -461,8 +455,8 @@ void broker_host_status(int type, int flags, int attr, host *hst)
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
-	hst->last_update = ds.timestamp.tv_sec;
+	tv_set(&ds.timestamp);
+	tv_clone(&hst->last_update, &ds.timestamp);
 
 	ds.object_ptr = (void *)hst;
 
@@ -485,8 +479,8 @@ void broker_service_status(int type, int flags, int attr, service *svc)
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
-	svc->last_update = ds.timestamp.tv_sec;
+	tv_set(&ds.timestamp);
+	tv_clone(&svc->last_update, &ds.timestamp);
 
 	ds.object_ptr = (void *)svc;
 
@@ -509,7 +503,7 @@ void broker_contact_status(int type, int flags, int attr, contact *cntct)
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.object_ptr = (void *)cntct;
 
@@ -534,7 +528,7 @@ neb_cb_resultset *broker_notification_data(int type, int flags, int attr, int no
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.notification_type = notification_type;
 	ds.start_time = start_time;
@@ -578,7 +572,7 @@ int broker_contact_notification_data(int type, int flags, int attr, int notifica
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.notification_type = notification_type;
 	ds.start_time = start_time;
@@ -638,7 +632,7 @@ int broker_contact_notification_method_data(int type, int flags, int attr, int n
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.notification_type = notification_type;
 	ds.start_time = start_time;
@@ -689,7 +683,7 @@ void broker_adaptive_program_data(int type, int flags, int attr, int command_typ
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.command_type = command_type;
 	ds.modified_host_attribute = modhattr;
@@ -716,7 +710,7 @@ void broker_adaptive_host_data(int type, int flags, int attr, host *hst, int com
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.command_type = command_type;
 	ds.modified_attribute = modattr;
@@ -742,7 +736,7 @@ void broker_adaptive_service_data(int type, int flags, int attr, service *svc, i
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.command_type = command_type;
 	ds.modified_attribute = modattr;
@@ -768,7 +762,7 @@ void broker_adaptive_contact_data(int type, int flags, int attr, contact *cntct,
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.command_type = command_type;
 	ds.modified_attribute = modattr;
@@ -798,7 +792,7 @@ int broker_external_command(int type, int flags, int attr, int command_type, tim
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.command_type = command_type;
 	ds.entry_time = entry_time;
@@ -822,7 +816,7 @@ void broker_aggregated_status_data(int type, int flags, int attr)
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	/* make callbacks */
 	neb_make_callbacks(NEBCALLBACK_AGGREGATED_STATUS_DATA, (void *)&ds);
@@ -843,7 +837,7 @@ void broker_retention_data(int type, int flags, int attr)
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	/* make callbacks */
 	neb_make_callbacks(NEBCALLBACK_RETENTION_DATA, (void *)&ds);
@@ -866,7 +860,7 @@ void broker_acknowledgement_data(int type, int flags, int attr, int acknowledgem
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.acknowledgement_type = acknowledgement_type;
 	if (acknowledgement_type == SERVICE_ACKNOWLEDGEMENT) {
@@ -909,7 +903,7 @@ void broker_statechange_data(int type, int flags, int attr, int statechange_type
 	ds.type = type;
 	ds.flags = flags;
 	ds.attr = attr;
-	get_broker_timestamp(&ds.timestamp);
+	tv_set(&ds.timestamp);
 
 	ds.statechange_type = statechange_type;
 	if (statechange_type == SERVICE_STATECHANGE) {
diff --git a/src/naemon/checks_host.c b/src/naemon/checks_host.c
index e57d84b92..89829179a 100644
--- a/src/naemon/checks_host.c
+++ b/src/naemon/checks_host.c
@@ -106,10 +106,11 @@ void checks_init_hosts(void)
 
 void schedule_next_host_check(host *hst, time_t delay, int options)
 {
-	time_t current_time = time(NULL);
+	struct timeval current_time;
+	tv_set(&current_time);
 
 	/* A closer check is already scheduled, skip this scheduling */
-	if (hst->next_check_event != NULL && hst->next_check < delay + current_time) {
+	if (hst->next_check_event != NULL && hst->next_check < delay + current_time.tv_sec) {
 		/*... unless this is a forced check or postponement is allowed*/
 		if (!(options & (CHECK_OPTION_FORCE_EXECUTION | CHECK_OPTION_ALLOW_POSTPONE)))
 			return;
@@ -122,8 +123,8 @@ void schedule_next_host_check(host *hst, time_t delay, int options)
 
 	/* Schedule the event */
 	hst->check_options = options;
-	hst->next_check = delay + current_time;
-	hst->last_update = current_time;
+	hst->next_check = delay + current_time.tv_sec;
+	tv_set(&hst->last_update);
 	hst->next_check_event = schedule_event(delay, handle_host_check_event, (void *)hst);
 
 	/* update the status log, since next_check and check_options is updated */
@@ -148,7 +149,7 @@ static void handle_host_check_event(struct nm_event_execution_properties *evprop
 	if (evprop->execution_type == EVENT_EXEC_NORMAL) {
 		/* get event latency */
 		latency = evprop->attributes.timed.latency;
-		gettimeofday(&tv, NULL);
+		tv_set(&tv);
 
 		/* When the callback is called, the pointer to the timed event is invalid */
 		hst->next_check_event = NULL;
@@ -208,7 +209,8 @@ static int run_async_host_check(host *hst, int check_options, double latency)
 	int runchk_result = OK;
 	int macro_options = STRIP_ILLEGAL_MACRO_CHARS | ESCAPE_MACRO_CHARS;
 	int neb_result = OK;
-	time_t now = time(NULL);
+	struct timeval now;
+	tv_set(&now);
 
 	log_debug_info(DEBUGL_CHECKS, 0, "** Running async check of host '%s'...\n", hst->name);
 
@@ -225,7 +227,7 @@ static int run_async_host_check(host *hst, int check_options, double latency)
 		}
 
 		/* abort if check was recently completed */
-		if (hst->last_check + cached_host_check_horizon > now && hst->last_check <= now) {
+		if (hst->last_check + cached_host_check_horizon > now.tv_sec && hst->last_check <= now.tv_sec) {
 			log_debug_info(DEBUGL_CHECKS, 0, "Host '%s' was last checked within its cache horizon. Aborting check\n", hst->name);
 			return ERROR;
 		}
@@ -277,7 +279,7 @@ static int run_async_host_check(host *hst, int check_options, double latency)
 	start_time.tv_usec = 0L;
 	end_time.tv_sec = 0L;
 	end_time.tv_usec = 0L;
-	hst->last_update = now;
+	tv_set(&hst->last_update);
 
 	neb_result = broker_host_check(NEBTYPE_HOSTCHECK_ASYNC_PRECHECK, NEBFLAG_NONE, NEBATTR_NONE, hst, CHECK_TYPE_ACTIVE, hst->current_state, hst->state_type, start_time, end_time, hst->check_command, hst->latency, 0.0, host_check_timeout, FALSE, 0, NULL, NULL, NULL, NULL, NULL);
 
@@ -321,7 +323,7 @@ static int run_async_host_check(host *hst, int check_options, double latency)
 	}
 
 	/* get the command start time */
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	cr = nm_calloc(1, sizeof(*cr));
 	init_check_result(cr);
@@ -385,7 +387,6 @@ int update_host_state_post_check(struct host *hst, struct check_result *cr)
 {
 	int result;
 	char *temp_ptr = NULL;
-	time_t now = time(NULL);
 
 	if (!hst || !cr)
 		return ERROR;
@@ -426,7 +427,7 @@ int update_host_state_post_check(struct host *hst, struct check_result *cr)
 
 	/* get the last check time */
 	hst->last_check = cr->start_time.tv_sec;
-	hst->last_update = now;
+	tv_set(&hst->last_update);
 
 	/* save the old host state */
 	hst->last_state = hst->current_state;
@@ -563,7 +564,6 @@ int update_host_state_post_check(struct host *hst, struct check_result *cr)
 int handle_async_host_check_result(host *temp_host, check_result *cr)
 {
 	int alert_recorded = NEBATTR_NONE;
-	struct timeval end_time_hires;
 	struct host pre;
 	int first_recorded_state = NEBATTR_NONE;
 
@@ -610,9 +610,6 @@ int handle_async_host_check_result(host *temp_host, check_result *cr)
 
 	log_debug_info(DEBUGL_CHECKS, 1, "** Async check result for host '%s' handled: new state=%d\n", temp_host->name, temp_host->current_state);
 
-	/* high resolution end time for event broker */
-	gettimeofday(&end_time_hires, NULL);
-
 	broker_host_check(
 	    NEBTYPE_HOSTCHECK_PROCESSED,
 	    NEBFLAG_NONE,
@@ -635,7 +632,7 @@ int handle_async_host_check_result(host *temp_host, check_result *cr)
 	    temp_host->perf_data,
 	    cr);
 
-	temp_host->last_update = end_time_hires.tv_sec;
+	tv_set(&temp_host->last_update);
 	return OK;
 }
 
@@ -643,7 +640,6 @@ static void handle_worker_host_check(wproc_result *wpres, void *arg, int flags)
 {
 	check_result *cr = (check_result *)arg;
 	struct host *hst;
-	time_t now = time(NULL);
 
 	/* decrement the number of host checks still out there... */
 	if (currently_running_host_checks > 0)
@@ -653,7 +649,7 @@ static void handle_worker_host_check(wproc_result *wpres, void *arg, int flags)
 		hst = find_host(cr->host_name);
 		if (hst) {
 			hst->is_executing = FALSE;
-			hst->last_update = now;
+			tv_set(&hst->last_update);
 			memcpy(&cr->rusage, &wpres->rusage, sizeof(wpres->rusage));
 			cr->start_time.tv_sec = wpres->start.tv_sec;
 			cr->start_time.tv_usec = wpres->start.tv_usec;
@@ -1161,14 +1157,14 @@ static void check_host_result_freshness(struct nm_event_execution_properties *ev
 static void check_for_orphaned_hosts_eventhandler(struct nm_event_execution_properties *evprop)
 {
 	host *temp_host = NULL;
-	time_t current_time = 0L;
 	time_t expected_time = 0L;
+	struct timeval current_time;
 
 	if (evprop->execution_type == EVENT_EXEC_NORMAL) {
 		schedule_event(DEFAULT_ORPHAN_CHECK_INTERVAL, check_for_orphaned_hosts_eventhandler, evprop->user_data);
 
 		/* get the current time */
-		time(&current_time);
+		tv_set(&current_time);
 
 		/* check all hosts... */
 		for (temp_host = host_list; temp_host != NULL; temp_host = temp_host->next) {
@@ -1185,7 +1181,7 @@ static void check_for_orphaned_hosts_eventhandler(struct nm_event_execution_prop
 			expected_time = (time_t)(temp_host->next_check + temp_host->latency + host_check_timeout + check_reaper_interval + 600);
 
 			/* this host was supposed to have executed a while ago, but for some reason the results haven't come back in... */
-			if (expected_time < current_time) {
+			if (expected_time < current_time.tv_sec) {
 
 				/* log a warning */
 				nm_log(NSLOG_RUNTIME_WARNING,
@@ -1200,7 +1196,7 @@ static void check_for_orphaned_hosts_eventhandler(struct nm_event_execution_prop
 				/* disable the executing flag */
 				temp_host->is_executing = FALSE;
 
-				temp_host->last_update = current_time;
+				tv_set(&temp_host->last_update);
 
 				/* schedule an immediate check of the host */
 				schedule_next_host_check(temp_host, 0, CHECK_OPTION_ORPHAN_CHECK);
diff --git a/src/naemon/checks_service.c b/src/naemon/checks_service.c
index baa220a69..f49c7dfed 100644
--- a/src/naemon/checks_service.c
+++ b/src/naemon/checks_service.c
@@ -107,10 +107,11 @@ void checks_init_services(void)
 
 void schedule_next_service_check(service *svc, time_t delay, int options)
 {
-	time_t current_time = time(NULL);
+	struct timeval current_time;
+	tv_set(&current_time);
 
 	/* A closer check is already scheduled, skip this scheduling */
-	if (svc->next_check_event != NULL && svc->next_check < delay + current_time) {
+	if (svc->next_check_event != NULL && svc->next_check < delay + current_time.tv_sec) {
 		/*... unless this is a forced check or postponement is allowed*/
 		if (!(options & (CHECK_OPTION_FORCE_EXECUTION | CHECK_OPTION_ALLOW_POSTPONE))) {
 			return;
@@ -124,8 +125,8 @@ void schedule_next_service_check(service *svc, time_t delay, int options)
 
 	/* Schedule the event */
 	svc->check_options = options;
-	svc->next_check = delay + current_time;
-	svc->last_update = current_time;
+	svc->next_check = delay + current_time.tv_sec;
+	tv_set(&svc->last_update);
 	svc->next_check_event = schedule_event(delay, handle_service_check_event, (void *)svc);
 
 	/* update the status log, since next_check and check_options is updated */
@@ -153,7 +154,7 @@ static void handle_service_check_event(struct nm_event_execution_properties *evp
 
 		/* get event latency */
 		latency = evprop->attributes.timed.latency;
-		gettimeofday(&tv, NULL);
+		tv_set(&tv);
 
 		/* When the callback is called, the pointer to the timed event is invalid */
 		temp_service->next_check_event = NULL;
@@ -352,8 +353,8 @@ static int run_scheduled_service_check(service *svc, int check_options, double l
 	}
 
 	/* get the command start time */
-	gettimeofday(&start_time, NULL);
-	svc->last_update = start_time.tv_sec;
+	tv_set(&start_time);
+	tv_clone(&svc->last_update, &start_time);
 
 	/* neb module wants to cancel the service check - the check will be rescheduled for a later time by the scheduling logic */
 	if (neb_result == NEBERROR_CALLBACKCANCEL) {
@@ -485,7 +486,6 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 	int hard_state_change = FALSE;
 	int first_host_check_initiated = FALSE;
 	int route_result = STATE_UP;
-	time_t current_time = 0L;
 	int alert_recorded = NEBATTR_NONE;
 	int first_recorded_state = NEBATTR_NONE;
 	char *old_plugin_output = NULL;
@@ -495,13 +495,14 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 	service *master_service = NULL;
 	int state_changes_use_cached_state = TRUE; /* TODO - 09/23/07 move this to a global variable */
 	int flapping_check_done = FALSE;
+	struct timeval current_time;
 
 	/* make sure we have what we need */
 	if (temp_service == NULL || queued_check_result == NULL)
 		return ERROR;
 
 	/* get the current time */
-	time(&current_time);
+	tv_set(&current_time);
 
 	log_debug_info(DEBUGL_CHECKS, 0, "** Handling check result for service '%s' on host '%s' from '%s'...\n", temp_service->description, temp_service->host_name, check_result_source(queued_check_result));
 	log_debug_info(DEBUGL_CHECKS, 1, "HOST: %s, SERVICE: %s, CHECK TYPE: %s, OPTIONS: %d, SCHEDULED: %s, EXITED OK: %s, RETURN CODE: %d, OUTPUT: %s\n", temp_service->host_name, temp_service->description, (queued_check_result->check_type == CHECK_TYPE_ACTIVE) ? "Active" : "Passive", queued_check_result->check_options, (queued_check_result->scheduled_check == TRUE) ? "Yes" : "No", (queued_check_result->exited_ok == TRUE) ? "Yes" : "No", queued_check_result->return_code, queued_check_result->output);
@@ -522,7 +523,7 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 		}
 	}
 
-	temp_service->last_update = current_time;
+	tv_set(&temp_service->last_update);
 
 	/* clear the freshening flag (it would have been set if this service was determined to be stale) */
 	if (queued_check_result->check_options & CHECK_OPTION_FRESHNESS_CHECK)
@@ -536,7 +537,7 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 	/* If a services goes stale, Nagios will initiate a forced check in order to freshen it.  There is a race condition whereby a passive check
 	   could arrive between the 1) initiation of the forced check and 2) the time when the forced check result is processed here.  This would
 	   make the service fresh again, so we do a quick check to make sure the service is still stale before we accept the check result. */
-	if ((queued_check_result->check_options & CHECK_OPTION_FRESHNESS_CHECK) && is_service_result_fresh(temp_service, current_time, FALSE) == TRUE) {
+	if ((queued_check_result->check_options & CHECK_OPTION_FRESHNESS_CHECK) && is_service_result_fresh(temp_service, current_time.tv_sec, FALSE) == TRUE) {
 		log_debug_info(DEBUGL_CHECKS, 0, "Discarding service freshness check result because the service is currently fresh (race condition avoided).\n");
 		return OK;
 	}
@@ -658,11 +659,11 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 
 		/* if the host has never been checked before, verify its status */
 		/* only do this if 1) the initial state was set to non-UP or 2) the host is not scheduled to be checked soon (next 5 minutes) */
-		if (temp_host->has_been_checked == FALSE && (temp_host->initial_state != STATE_UP || (unsigned long)temp_host->next_check == 0L || (unsigned long)(temp_host->next_check - current_time) > 300)) {
+		if (temp_host->has_been_checked == FALSE && (temp_host->initial_state != STATE_UP || (unsigned long)temp_host->next_check == 0L || (unsigned long)(temp_host->next_check - current_time.tv_sec) > 300)) {
 
 			/* set a flag to remember that we launched a check */
 			first_host_check_initiated = TRUE;
-			schedule_host_check(temp_host, current_time, CHECK_OPTION_DEPENDENCY_CHECK);
+			schedule_host_check(temp_host, current_time.tv_sec, CHECK_OPTION_DEPENDENCY_CHECK);
 		}
 	}
 
@@ -758,7 +759,7 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 			/* don't reset last problem id, or it will be zero the next time a problem is encountered */
 			nm_free(temp_service->current_problem_id);
 			temp_service->current_problem_id = (char*)g_uuid_string_random();
-			temp_service->problem_start = current_time;
+			temp_service->problem_start = current_time.tv_sec;
 			temp_service->problem_end = 0L;
 		}
 
@@ -768,7 +769,7 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 			temp_service->last_problem_id = temp_service->current_problem_id;
 			temp_service->current_problem_id = NULL;
 			if(temp_service->problem_start > 0)
-				temp_service->problem_end = current_time;
+				temp_service->problem_end = current_time.tv_sec;
 		}
 	}
 
@@ -797,15 +798,15 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 			else {
 				/* can we use the last cached host state? */
 				/* usually only use cached host state if no service state change has occurred */
-				if ((state_change == FALSE || state_changes_use_cached_state == TRUE) && temp_host->has_been_checked == TRUE && (temp_host->last_check + cached_host_check_horizon > current_time && temp_host->last_check <= current_time)) {
+				if ((state_change == FALSE || state_changes_use_cached_state == TRUE) && temp_host->has_been_checked == TRUE && (temp_host->last_check + cached_host_check_horizon > current_time.tv_sec && temp_host->last_check <= current_time.tv_sec)) {
 					log_debug_info(DEBUGL_CHECKS, 1, "* Using cached host state: %d\n", temp_host->current_state);
-					update_check_stats(ACTIVE_ONDEMAND_HOST_CHECK_STATS, current_time);
-					update_check_stats(ACTIVE_CACHED_HOST_CHECK_STATS, current_time);
+					update_check_stats(ACTIVE_ONDEMAND_HOST_CHECK_STATS, current_time.tv_sec);
+					update_check_stats(ACTIVE_CACHED_HOST_CHECK_STATS, current_time.tv_sec);
 				}
 
 				/* else launch an async (parallel) check of the host */
 				else
-					schedule_host_check(temp_host, current_time, CHECK_OPTION_DEPENDENCY_CHECK);
+					schedule_host_check(temp_host, current_time.tv_sec, CHECK_OPTION_DEPENDENCY_CHECK);
 			}
 		}
 
@@ -897,10 +898,10 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 			log_debug_info(DEBUGL_CHECKS, 1, "Host is currently UP, so we'll recheck its state to make sure...\n");
 
 			/* only run a new check if we can and have to */
-			if (!execute_host_checks || (temp_host->last_check + cached_host_check_horizon > current_time && temp_host->last_check <= current_time)) {
+			if (!execute_host_checks || (temp_host->last_check + cached_host_check_horizon > current_time.tv_sec && temp_host->last_check <= current_time.tv_sec)) {
 				log_debug_info(DEBUGL_CHECKS, 1, "* Using cached host state: %d\n", temp_host->current_state);
-				update_check_stats(ACTIVE_ONDEMAND_HOST_CHECK_STATS, current_time);
-				update_check_stats(ACTIVE_CACHED_HOST_CHECK_STATS, current_time);
+				update_check_stats(ACTIVE_ONDEMAND_HOST_CHECK_STATS, current_time.tv_sec);
+				update_check_stats(ACTIVE_CACHED_HOST_CHECK_STATS, current_time.tv_sec);
 			} else if (state_change || temp_service->state_type == SOFT_STATE) {
 				schedule_next_host_check(temp_host, 0, CHECK_OPTION_DEPENDENCY_CHECK);
 			}
@@ -912,7 +913,7 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 			log_debug_info(DEBUGL_CHECKS, 1, "Host is currently %s.\n", host_state_name(temp_host->current_state));
 
 			if (execute_host_checks && (state_change || temp_service->state_type == SOFT_STATE)) {
-				schedule_host_check(temp_host, current_time, CHECK_OPTION_NONE);
+				schedule_host_check(temp_host, current_time.tv_sec, CHECK_OPTION_NONE);
 			}
 			/* else fake the host check, but (possibly) resend host notifications to contacts... */
 			else {
@@ -1036,7 +1037,7 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 					if (temp_dependency->dependent_service_ptr == temp_service && temp_dependency->master_service_ptr != NULL) {
 						master_service = (service *)temp_dependency->master_service_ptr;
 						log_debug_info(DEBUGL_CHECKS, 2, "Predictive check of service '%s' on host '%s' queued.\n", master_service->description, master_service->host_name);
-						schedule_service_check(master_service, current_time, CHECK_OPTION_DEPENDENCY_CHECK);
+						schedule_service_check(master_service, current_time.tv_sec, CHECK_OPTION_DEPENDENCY_CHECK);
 					}
 				}
 				for (list = temp_service->notify_deps; list; list = list->next) {
@@ -1044,7 +1045,7 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 					if (temp_dependency->dependent_service_ptr == temp_service && temp_dependency->master_service_ptr != NULL) {
 						master_service = (service *)temp_dependency->master_service_ptr;
 						log_debug_info(DEBUGL_CHECKS, 2, "Predictive check of service '%s' on host '%s' queued.\n", master_service->description, master_service->host_name);
-						schedule_service_check(master_service, current_time, CHECK_OPTION_DEPENDENCY_CHECK);
+						schedule_service_check(master_service, current_time.tv_sec, CHECK_OPTION_DEPENDENCY_CHECK);
 					}
 				}
 			}
@@ -1167,15 +1168,15 @@ int handle_async_service_check_result(service *temp_service, check_result *queue
 static void check_for_orphaned_services_eventhandler(struct nm_event_execution_properties *evprop)
 {
 	service *temp_service = NULL;
-	time_t current_time = 0L;
 	time_t expected_time = 0L;
+	struct timeval current_time;
 
 	if (evprop->execution_type == EVENT_EXEC_NORMAL) {
 
 		schedule_event(DEFAULT_ORPHAN_CHECK_INTERVAL, check_for_orphaned_services_eventhandler, evprop->user_data);
 
 		/* get the current time */
-		time(&current_time);
+		tv_set(&current_time);
 
 		/* check all services... */
 		for (temp_service = service_list; temp_service != NULL; temp_service = temp_service->next) {
@@ -1188,7 +1189,7 @@ static void check_for_orphaned_services_eventhandler(struct nm_event_execution_p
 			expected_time = (time_t)(temp_service->next_check + temp_service->latency + service_check_timeout + check_reaper_interval + 600);
 
 			/* this service was supposed to have executed a while ago, but for some reason the results haven't come back in... */
-			if (expected_time < current_time) {
+			if (expected_time < current_time.tv_sec) {
 
 				/* log a warning */
 				nm_log(NSLOG_RUNTIME_WARNING,
@@ -1205,7 +1206,7 @@ static void check_for_orphaned_services_eventhandler(struct nm_event_execution_p
 
 				/* disable the executing flag */
 				temp_service->is_executing = FALSE;
-				temp_service->last_update = current_time;
+				tv_set(&temp_service->last_update);
 
 				/* schedule an immediate check of the service */
 				schedule_next_service_check(temp_service, 0, CHECK_OPTION_ORPHAN_CHECK);
diff --git a/src/naemon/commands.c b/src/naemon/commands.c
index de2003514..54dcce30d 100644
--- a/src/naemon/commands.c
+++ b/src/naemon/commands.c
@@ -1832,14 +1832,11 @@ static int host_command_handler(const struct external_command *ext_command, time
 	unsigned long downtime_id = 0L;
 	unsigned long duration = 0L;
 	time_t old_interval = 0L;
-	time_t current_time = 0L;
 
 	if (ext_command->id != CMD_DEL_HOST_COMMENT) {
 		target_host = GV("host_name");
-		if (target_host) {
-			time(&current_time);
-			target_host->last_update = current_time;
-		}
+		if (target_host)
+			tv_set(&target_host->last_update);
 	}
 
 	switch (ext_command->id) {
@@ -2237,15 +2234,12 @@ static int service_command_handler(const struct external_command *ext_command, t
 {
 	struct service *target_service = NULL;
 	time_t old_interval = 0L;
-	time_t current_time = 0L;
 	unsigned long duration = 0L;
 
 	if (ext_command->id != CMD_DEL_SVC_COMMENT) {
 		target_service = GV_SERVICE("service");
-		if (target_service) {
-			time(&current_time);
-			target_service->last_update = current_time;
-		}
+		if (target_service)
+			tv_set(&target_service->last_update);
 	}
 
 	switch (ext_command->id) {
@@ -2558,7 +2552,6 @@ static int contactgroup_command_handler(const struct external_command *ext_comma
 
 static int change_custom_var_handler(const struct external_command *ext_command, time_t entry_time)
 {
-	time_t current_time = 0L;
 	customvariablesmember *customvariablesmember_p = NULL;
 	char *varname;
 	int x = 0;
@@ -2566,18 +2559,16 @@ static int change_custom_var_handler(const struct external_command *ext_command,
 	struct host * target_host = NULL;
 	struct contact * target_contact = NULL;
 
-	time(&current_time);
-
 	switch (ext_command->id) {
 	case CMD_CHANGE_CUSTOM_SVC_VAR:
 		target_service = GV_SERVICE("service");
-		target_service->last_update = current_time;
+		tv_set(&target_service->last_update);
 		customvariablesmember_p = target_service->custom_variables;
 		break;
 
 	case CMD_CHANGE_CUSTOM_HOST_VAR:
 		target_host = GV_HOST("host_name");
-		target_host->last_update = current_time;
+		tv_set(&target_host->last_update);
 		customvariablesmember_p = target_host->custom_variables;
 		break;
 
@@ -3537,7 +3528,7 @@ int process_passive_service_check(time_t check_time, char *host_name, char *svc_
 		cr.return_code = STATE_UNKNOWN;
 
 	/* calculate latency */
-	gettimeofday(&tv, NULL);
+	tv_set(&tv);
 	cr.latency = (double)((double)(tv.tv_sec - check_time) + (double)(tv.tv_usec / 1000.0) / 1000.0);
 	if (cr.latency < 0.0)
 		cr.latency = 0.0;
@@ -3587,7 +3578,7 @@ int process_passive_host_check(time_t check_time, char *host_name, int return_co
 	cr.return_code = return_code;
 
 	/* calculate latency */
-	gettimeofday(&tv, NULL);
+	tv_set(&tv);
 	cr.latency = (double)((double)(tv.tv_sec - check_time) + (double)(tv.tv_usec / 1000.0) / 1000.0);
 	if (cr.latency < 0.0)
 		cr.latency = 0.0;
diff --git a/src/naemon/logging.c b/src/naemon/logging.c
index 16ac9ae0a..b22005ebb 100644
--- a/src/naemon/logging.c
+++ b/src/naemon/logging.c
@@ -323,7 +323,7 @@ int log_debug_info(int level, int verbosity, const char *fmt, ...)
 		return ERROR;
 
 	/* write the timestamp */
-	gettimeofday(&current_time, NULL);
+	tv_set(&current_time);
 	fprintf(debug_file_fp, "[%ld.%06ld] [%03d.%d] [pid=%lu] ", (long)current_time.tv_sec, (long)current_time.tv_usec, level, verbosity, (unsigned long)getpid());
 
 	/* write the data */
diff --git a/src/naemon/notifications.c b/src/naemon/notifications.c
index e2e6cb432..13b21e66e 100644
--- a/src/naemon/notifications.c
+++ b/src/naemon/notifications.c
@@ -347,7 +347,6 @@ int service_notification(service *svc, int type, char *not_author, char *not_dat
 	notification *notification_list = NULL;
 	notification *temp_notification = NULL;
 	contact *temp_contact = NULL;
-	time_t current_time;
 	struct timeval start_time;
 	struct timeval end_time;
 	int escalated = FALSE;
@@ -360,8 +359,7 @@ int service_notification(service *svc, int type, char *not_author, char *not_dat
 	neb_cb_result *neb_result = NULL;
 
 	/* get the current time */
-	time(&current_time);
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	log_debug_info(DEBUGL_NOTIFICATIONS, 0, "** Service Notification Attempt ** Host: '%s', Service: '%s', Type: %s, Options: %d, Current State: %d, Last Notification: %s\n", svc->host_name, svc->description, notification_reason_name(type), options, svc->current_state, ctime(&svc->last_notification));
 
@@ -537,16 +535,16 @@ int service_notification(service *svc, int type, char *not_author, char *not_dat
 			if (contacts_notified > 0) {
 
 				/* calculate the next acceptable re-notification time */
-				svc->next_notification = get_next_service_notification_time(svc, current_time);
+				svc->next_notification = get_next_service_notification_time(svc, start_time.tv_sec);
 
 				log_debug_info(DEBUGL_NOTIFICATIONS, 0, "%d contacts were notified.  Next possible notification time: %s\n", contacts_notified, ctime(&svc->next_notification));
 
 				/* update the last notification time for this service (this is needed for rescheduling later notifications) */
-				svc->last_notification = current_time;
+				svc->last_notification = start_time.tv_sec;
 
 				/* update notifications flags */
 				add_notified_on(svc, svc->current_state);
-				svc->last_update = current_time;
+				tv_set(&svc->last_update);
 			}
 
 			/* we didn't end up notifying anyone */
@@ -556,7 +554,7 @@ int service_notification(service *svc, int type, char *not_author, char *not_dat
 				svc->current_notification_number--;
 
 				log_debug_info(DEBUGL_NOTIFICATIONS, 0, "No contacts were notified.  Next possible notification time: %s\n", ctime(&svc->next_notification));
-				svc->last_update = current_time;
+				tv_set(&svc->last_update);
 			}
 		}
 
@@ -577,7 +575,7 @@ int service_notification(service *svc, int type, char *not_author, char *not_dat
 	nm_free(mac.x[MACRO_NOTIFICATIONISESCALATED]);
 
 	/* get the time we finished */
-	gettimeofday(&end_time, NULL);
+	tv_set(&end_time);
 
 	neb_resultset = broker_notification_data(NEBTYPE_NOTIFICATION_END, NEBFLAG_NONE, NEBATTR_NONE, SERVICE_NOTIFICATION, type, start_time, end_time, (void *)svc, not_author, not_data, escalated, contacts_notified);
 	neb_cb_resultset_destroy(neb_resultset);
@@ -961,7 +959,7 @@ int notify_contact_of_service(nagios_macros *mac, contact *cntct, service *svc,
 	log_debug_info(DEBUGL_NOTIFICATIONS, 2, "** Notifying contact '%s'\n", cntct->name);
 
 	/* get start time */
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	end_time.tv_sec = 0L;
 	end_time.tv_usec = 0L;
@@ -975,7 +973,7 @@ int notify_contact_of_service(nagios_macros *mac, contact *cntct, service *svc,
 	for (temp_commandsmember = cntct->service_notification_commands; temp_commandsmember != NULL; temp_commandsmember = temp_commandsmember->next) {
 
 		/* get start time */
-		gettimeofday(&method_start_time, NULL);
+		tv_set(&method_start_time);
 
 		method_end_time.tv_sec = 0L;
 		method_end_time.tv_usec = 0L;
@@ -1033,13 +1031,13 @@ int notify_contact_of_service(nagios_macros *mac, contact *cntct, service *svc,
 		nm_free(processed_command);
 
 		/* get end time */
-		gettimeofday(&method_end_time, NULL);
+		tv_set(&method_end_time);
 
 		broker_contact_notification_method_data(NEBTYPE_CONTACTNOTIFICATIONMETHOD_END, NEBFLAG_NONE, NEBATTR_NONE, SERVICE_NOTIFICATION, type, method_start_time, method_end_time, (void *)svc, cntct, temp_commandsmember->command, not_author, not_data, escalated);
 	}
 
 	/* get end time */
-	gettimeofday(&end_time, NULL);
+	tv_set(&end_time);
 
 	/* update the contact's last service notification time */
 	cntct->last_service_notification = start_time.tv_sec;
@@ -1236,7 +1234,6 @@ int host_notification(host *hst, int type, char *not_author, char *not_data, int
 	notification *notification_list = NULL;
 	notification *temp_notification = NULL;
 	contact *temp_contact = NULL;
-	time_t current_time;
 	struct timeval start_time;
 	struct timeval end_time;
 	int escalated = FALSE;
@@ -1249,8 +1246,7 @@ int host_notification(host *hst, int type, char *not_author, char *not_data, int
 	neb_cb_result *neb_result = NULL;
 
 	/* get the current time */
-	time(&current_time);
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	log_debug_info(DEBUGL_NOTIFICATIONS, 0, "** Host Notification Attempt ** Host: '%s', Type: %s, Options: %d, Current State: %d, Last Notification: %s\n", hst->name, notification_reason_name(type), options, hst->current_state, ctime(&hst->last_notification));
 
@@ -1422,16 +1418,16 @@ int host_notification(host *hst, int type, char *not_author, char *not_data, int
 			if (contacts_notified > 0) {
 
 				/* calculate the next acceptable re-notification time */
-				hst->next_notification = get_next_host_notification_time(hst, current_time);
+				hst->next_notification = get_next_host_notification_time(hst, start_time.tv_sec);
 
 				/* update the last notification time for this host (this is needed for scheduling the next problem notification) */
-				hst->last_notification = current_time;
+				hst->last_notification = start_time.tv_sec;
 
 				/* update notifications flags */
 				add_notified_on(hst, hst->current_state);
 
 				log_debug_info(DEBUGL_NOTIFICATIONS, 0, "%d contacts were notified.  Next possible notification time: %s\n", contacts_notified, ctime(&hst->next_notification));
-				hst->last_update = current_time;
+				tv_set(&hst->last_update);
 			}
 
 			/* we didn't end up notifying anyone */
@@ -1441,7 +1437,7 @@ int host_notification(host *hst, int type, char *not_author, char *not_data, int
 				hst->current_notification_number--;
 
 				log_debug_info(DEBUGL_NOTIFICATIONS, 0, "No contacts were notified.  Next possible notification time: %s\n", ctime(&hst->next_notification));
-				hst->last_update = current_time;
+				tv_set(&hst->last_update);
 			}
 		}
 
@@ -1461,7 +1457,7 @@ int host_notification(host *hst, int type, char *not_author, char *not_data, int
 	nm_free(mac.x[MACRO_NOTIFICATIONISESCALATED]);
 
 	/* get the time we finished */
-	gettimeofday(&end_time, NULL);
+	tv_set(&end_time);
 
 	neb_resultset = broker_notification_data(NEBTYPE_NOTIFICATION_END, NEBFLAG_NONE, NEBATTR_NONE, HOST_NOTIFICATION, type, start_time, end_time, (void *)hst, not_author, not_data, escalated, contacts_notified);
 	neb_cb_resultset_destroy(neb_resultset);
@@ -1807,7 +1803,7 @@ int notify_contact_of_host(nagios_macros *mac, contact *cntct, host *hst, int ty
 	log_debug_info(DEBUGL_NOTIFICATIONS, 2, "** Notifying contact '%s'\n", cntct->name);
 
 	/* get start time */
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	end_time.tv_sec = 0L;
 	end_time.tv_usec = 0L;
@@ -1821,7 +1817,7 @@ int notify_contact_of_host(nagios_macros *mac, contact *cntct, host *hst, int ty
 	for (temp_commandsmember = cntct->host_notification_commands; temp_commandsmember != NULL; temp_commandsmember = temp_commandsmember->next) {
 
 		/* get start time */
-		gettimeofday(&method_start_time, NULL);
+		tv_set(&method_start_time);
 
 		method_end_time.tv_sec = 0L;
 		method_end_time.tv_usec = 0L;
@@ -1879,13 +1875,13 @@ int notify_contact_of_host(nagios_macros *mac, contact *cntct, host *hst, int ty
 		nm_free(processed_command);
 
 		/* get end time */
-		gettimeofday(&method_end_time, NULL);
+		tv_set(&method_end_time);
 
 		broker_contact_notification_method_data(NEBTYPE_CONTACTNOTIFICATIONMETHOD_END, NEBFLAG_NONE, NEBATTR_NONE, HOST_NOTIFICATION, type, method_start_time, method_end_time, (void *)hst, cntct, temp_commandsmember->command, not_author, not_data, escalated);
 	}
 
 	/* get end time */
-	gettimeofday(&end_time, NULL);
+	tv_set(&end_time);
 
 	/* update the contact's last host notification time */
 	cntct->last_host_notification = start_time.tv_sec;
diff --git a/src/naemon/objects_host.h b/src/naemon/objects_host.h
index 099bec320..68e2852cc 100644
--- a/src/naemon/objects_host.h
+++ b/src/naemon/objects_host.h
@@ -133,7 +133,7 @@ struct host {
 	/* objects we depend upon */
 	struct objectlist *exec_deps, *notify_deps;
 	struct objectlist *escalation_list;
-	time_t  last_update /* timestamp when object has been updated the last time */;
+	struct timeval  last_update /* timestamp when object has been updated the last time */;
 	struct  host *next;
 	struct timed_event *next_check_event;
 };
diff --git a/src/naemon/objects_service.h b/src/naemon/objects_service.h
index 35031b85d..3b2551107 100644
--- a/src/naemon/objects_service.h
+++ b/src/naemon/objects_service.h
@@ -124,7 +124,7 @@ struct service {
 	struct objectlist *servicegroups_ptr;
 	struct objectlist *exec_deps, *notify_deps;
 	struct objectlist *escalation_list;
-	time_t  last_update /* timestamp when object has been updated the last time */;
+	struct timeval last_update /* timestamp when object has been updated the last time */;
 	struct service *next;
 	struct timed_event *next_check_event;
 };
diff --git a/src/naemon/sehandlers.c b/src/naemon/sehandlers.c
index 27c1d81cf..8a67ace23 100644
--- a/src/naemon/sehandlers.c
+++ b/src/naemon/sehandlers.c
@@ -242,7 +242,7 @@ static int run_global_service_event_handler(nagios_macros *mac, service *svc)
 	log_debug_info(DEBUGL_EVENTHANDLERS, 1, "Running global event handler for service '%s' on host '%s'...\n", svc->description, svc->host_name);
 
 	/* get start time */
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	get_raw_command_line_r(mac, global_service_event_handler_ptr, global_service_event_handler, &raw_command, macro_options);
 	if (raw_command == NULL) {
@@ -285,7 +285,7 @@ static int run_global_service_event_handler(nagios_macros *mac, service *svc)
 		nm_log(NSLOG_EVENT_HANDLER | NSLOG_RUNTIME_WARNING, "Warning: Global service event handler command '%s' timed out after %d seconds\n", processed_command, event_handler_timeout);
 
 	/* get end time */
-	gettimeofday(&end_time, NULL);
+	tv_set(&end_time);
 
 	broker_event_handler(NEBTYPE_EVENTHANDLER_END, NEBFLAG_NONE, NEBATTR_NONE, GLOBAL_SERVICE_EVENTHANDLER, (void *)svc, svc->current_state, svc->state_type, start_time, end_time, exectime, event_handler_timeout, early_timeout, result, global_service_event_handler, processed_command, command_output);
 
@@ -324,7 +324,7 @@ static int run_service_event_handler(nagios_macros *mac, service *svc)
 	log_debug_info(DEBUGL_EVENTHANDLERS, 1, "Running event handler for service '%s' on host '%s'...\n", svc->description, svc->host_name);
 
 	/* get start time */
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	get_raw_command_line_r(mac, svc->event_handler_ptr, svc->event_handler, &raw_command, macro_options);
 	if (raw_command == NULL)
@@ -366,7 +366,7 @@ static int run_service_event_handler(nagios_macros *mac, service *svc)
 		nm_log(NSLOG_EVENT_HANDLER | NSLOG_RUNTIME_WARNING, "Warning: Service event handler command '%s' timed out after %d seconds\n", processed_command, event_handler_timeout);
 
 	/* get end time */
-	gettimeofday(&end_time, NULL);
+	tv_set(&end_time);
 
 	broker_event_handler(NEBTYPE_EVENTHANDLER_END, NEBFLAG_NONE, NEBATTR_NONE, SERVICE_EVENTHANDLER, (void *)svc, svc->current_state, svc->state_type, start_time, end_time, exectime, event_handler_timeout, early_timeout, result, svc->event_handler, processed_command, command_output);
 
@@ -444,7 +444,7 @@ static int run_global_host_event_handler(nagios_macros *mac, const host *const h
 	log_debug_info(DEBUGL_EVENTHANDLERS, 1, "Running global event handler for host '%s'..\n", hst->name);
 
 	/* get start time */
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	get_raw_command_line_r(mac, global_host_event_handler_ptr, global_host_event_handler, &raw_command, macro_options);
 	if (raw_command == NULL)
@@ -486,7 +486,7 @@ static int run_global_host_event_handler(nagios_macros *mac, const host *const h
 		nm_log(NSLOG_EVENT_HANDLER | NSLOG_RUNTIME_WARNING, "Warning: Global host event handler command '%s' timed out after %d seconds\n", processed_command, event_handler_timeout);
 
 	/* get end time */
-	gettimeofday(&end_time, NULL);
+	tv_set(&end_time);
 
 	broker_event_handler(NEBTYPE_EVENTHANDLER_END, NEBFLAG_NONE, NEBATTR_NONE, GLOBAL_HOST_EVENTHANDLER, (void *)hst, hst->current_state, hst->state_type, start_time, end_time, exectime, event_handler_timeout, early_timeout, result, global_host_event_handler, processed_command, command_output);
 
@@ -525,7 +525,7 @@ static int run_host_event_handler(nagios_macros *mac, const host *const hst)
 	log_debug_info(DEBUGL_EVENTHANDLERS, 1, "Running event handler for host '%s'..\n", hst->name);
 
 	/* get start time */
-	gettimeofday(&start_time, NULL);
+	tv_set(&start_time);
 
 	get_raw_command_line_r(mac, hst->event_handler_ptr, hst->event_handler, &raw_command, macro_options);
 	if (raw_command == NULL)
@@ -567,7 +567,7 @@ static int run_host_event_handler(nagios_macros *mac, const host *const hst)
 		nm_log(NSLOG_EVENT_HANDLER | NSLOG_RUNTIME_WARNING, "Warning: Host event handler command '%s' timed out after %d seconds\n", processed_command, event_handler_timeout);
 
 	/* get end time */
-	gettimeofday(&end_time, NULL);
+	tv_set(&end_time);
 
 	broker_event_handler(NEBTYPE_EVENTHANDLER_END, NEBFLAG_NONE, NEBATTR_NONE, HOST_EVENTHANDLER, (void *)hst, hst->current_state, hst->state_type, start_time, end_time, exectime, event_handler_timeout, early_timeout, result, hst->event_handler, processed_command, command_output);
 
diff --git a/src/naemon/shared.c b/src/naemon/shared.c
index 5e8160d44..ca35834be 100644
--- a/src/naemon/shared.c
+++ b/src/naemon/shared.c
@@ -68,15 +68,13 @@ void timing_point(const char *fmt, ...)
 		return;
 
 	if (first.tv_sec == 0) {
-		gettimeofday(&first, NULL);
-		last.tv_sec = first.tv_sec;
-		last.tv_usec = first.tv_usec;
+		tv_set(&first);
+		tv_clone(&last, &first);
 		printf("[0.0000 (+0.0000)] ");
 	} else {
-		gettimeofday(&now, NULL);
+		tv_set(&now);
 		printf("[%.4f (+%.4f)] ", tv_delta_f(&first, &now), tv_delta_f(&last, &now));
-		last.tv_sec = now.tv_sec;
-		last.tv_usec = now.tv_usec;
+		tv_clone(&last, &now);
 	}
 	va_start(ap, fmt);
 	vprintf(fmt, ap);
diff --git a/src/naemon/workers.c b/src/naemon/workers.c
index f15b98bef..d9a0ffbda 100644
--- a/src/naemon/workers.c
+++ b/src/naemon/workers.c
@@ -163,7 +163,7 @@ static void run_job_callback(struct wproc_job *job, struct wproc_result *wpres,
 {
 	if (!job || !job->callback)
 		return;
-	
+
 	if (!wpres) {
 		return;
 	}
@@ -290,22 +290,6 @@ void free_worker_memory(int flags)
 	workers.idx = 0;
 }
 
-static int str2timeval(char *str, struct timeval *tv)
-{
-	char *ptr, *ptr2;
-
-	tv->tv_sec = strtoul(str, &ptr, 10);
-	if (ptr == str) {
-		tv->tv_sec = tv->tv_usec = 0;
-		return -1;
-	}
-	if (*ptr == '.' || *ptr == ',') {
-		ptr2 = ptr + 1;
-		tv->tv_usec = strtoul(ptr2, &ptr, 10);
-	}
-	return 0;
-}
-
 /*
  * parses a worker result. We do no strdup()'s here, so when
  * kvv is destroyed, all references to strings will become
diff --git a/src/naemon/xrddefault.c b/src/naemon/xrddefault.c
index 2b0ead917..bf8258432 100644
--- a/src/naemon/xrddefault.c
+++ b/src/naemon/xrddefault.c
@@ -237,7 +237,7 @@ int xrddefault_save_state_information(void)
 		fprintf(fp, "is_flapping=%d\n", temp_host->is_flapping);
 		fprintf(fp, "percent_state_change=%.2f\n", temp_host->percent_state_change);
 		fprintf(fp, "check_flapping_recovery_notification=%d\n", temp_host->check_flapping_recovery_notification);
-		fprintf(fp, "last_update=%lu\n", temp_host->last_update);
+		fprintf(fp, "last_update=%s\n", tv_str(&temp_host->last_update));
 
 		fprintf(fp, "state_history=");
 		for (x = 0; x < MAX_STATE_HISTORY_ENTRIES; x++)
@@ -333,7 +333,7 @@ int xrddefault_save_state_information(void)
 			fprintf(fp, "config:obsess=%d\n", conf_svc->obsess);
 			fprintf(fp, "obsess=%d\n", temp_service->obsess);
 		}
-		fprintf(fp, "last_update=%lu\n", temp_service->last_update);
+		fprintf(fp, "last_update=%s\n", tv_str(&temp_service->last_update));
 		fprintf(fp, "is_flapping=%d\n", temp_service->is_flapping);
 		fprintf(fp, "percent_state_change=%.2f\n", temp_service->percent_state_change);
 		fprintf(fp, "check_flapping_recovery_notification=%d\n", temp_service->check_flapping_recovery_notification);
@@ -1095,7 +1095,7 @@ int xrddefault_read_state_information(void)
 						else if (!strcmp(var, "last_time_unreachable"))
 							temp_host->last_time_unreachable = strtoul(val, NULL, 10);
 						else if (!strcmp(var, "last_update"))
-							temp_host->last_update = strtoul(val, NULL, 10);
+							str2timeval(val, &temp_host->last_update);
 						else if (!strcmp(var, "notified_on_down"))
 							temp_host->notified_on |= (atoi(val) > 0 ? OPT_DOWN : 0);
 						else if (!strcmp(var, "notified_on_unreachable"))
@@ -1350,7 +1350,7 @@ int xrddefault_read_state_information(void)
 						else if (!strcmp(var, "last_time_critical"))
 							temp_service->last_time_critical = strtoul(val, NULL, 10);
 						else if (!strcmp(var, "last_update"))
-							temp_service->last_update = strtoul(val, NULL, 10);
+							str2timeval(val, &temp_service->last_update);
 						else if (!strcmp(var, "plugin_output")) {
 							nm_free(temp_service->plugin_output);
 							temp_service->plugin_output = nm_strdup(val);
diff --git a/src/naemon/xsddefault.c b/src/naemon/xsddefault.c
index 3eb5c1bca..980b2c3ba 100644
--- a/src/naemon/xsddefault.c
+++ b/src/naemon/xsddefault.c
@@ -233,11 +233,10 @@ int xsddefault_save_status_data(void)
 		fprintf(fp, "\tflap_detection_enabled=%d\n", temp_host->flap_detection_enabled);
 		fprintf(fp, "\tprocess_performance_data=%d\n", temp_host->process_performance_data);
 		fprintf(fp, "\tobsess=%d\n", temp_host->obsess);
-		fprintf(fp, "\tlast_update=%lu\n", current_time);
 		fprintf(fp, "\tis_flapping=%d\n", temp_host->is_flapping);
 		fprintf(fp, "\tpercent_state_change=%.2f\n", temp_host->percent_state_change);
 		fprintf(fp, "\tscheduled_downtime_depth=%d\n", temp_host->scheduled_downtime_depth);
-		fprintf(fp, "\tlast_update=%lu\n", temp_host->last_update);
+		fprintf(fp, "\tlast_update=%s\n", tv_str(&temp_host->last_update));
 		/* custom variables */
 		for (temp_customvariablesmember = temp_host->custom_variables; temp_customvariablesmember != NULL; temp_customvariablesmember = temp_customvariablesmember->next) {
 			if (temp_customvariablesmember->variable_name)
@@ -303,11 +302,10 @@ int xsddefault_save_status_data(void)
 		fprintf(fp, "\tflap_detection_enabled=%d\n", temp_service->flap_detection_enabled);
 		fprintf(fp, "\tprocess_performance_data=%d\n", temp_service->process_performance_data);
 		fprintf(fp, "\tobsess=%d\n", temp_service->obsess);
-		fprintf(fp, "\tlast_update=%lu\n", current_time);
 		fprintf(fp, "\tis_flapping=%d\n", temp_service->is_flapping);
 		fprintf(fp, "\tpercent_state_change=%.2f\n", temp_service->percent_state_change);
 		fprintf(fp, "\tscheduled_downtime_depth=%d\n", temp_service->scheduled_downtime_depth);
-		fprintf(fp, "\tlast_update=%lu\n", temp_service->last_update);
+		fprintf(fp, "\tlast_update=%s\n", tv_str(&temp_service->last_update));
 		/* custom variables */
 		for (temp_customvariablesmember = temp_service->custom_variables; temp_customvariablesmember != NULL; temp_customvariablesmember = temp_customvariablesmember->next) {
 			if (temp_customvariablesmember->variable_name)
diff --git a/src/worker/worker.c b/src/worker/worker.c
index 47ccbf79c..c31983f31 100644
--- a/src/worker/worker.c
+++ b/src/worker/worker.c
@@ -184,7 +184,7 @@ static int finish_job(child_process *cp, int reason)
 		exit_worker(1, "Failed to init response key/value vector");
 	}
 
-	gettimeofday(&cp->ei->stop, NULL);
+	tv_set(&cp->ei->stop);
 
 	cp->ei->runtime = tv_delta_f(&cp->ei->start, &cp->ei->stop);
 
@@ -561,7 +561,7 @@ static void spawn_job(struct kvvec *kvv)
 		return;
 	}
 
-	gettimeofday(&cp->ei->start, NULL);
+	tv_set(&cp->ei->start);
 	cp->request = kvv;
 	cp->ei->timed_event = schedule_event(cp->timeout, kill_job, cp);
 	cp->outstd.buf = nm_bufferqueue_create();
